### PR TITLE
(bugfix/kubernetes/rbac): Bring alpha and beta rbac rules into sync

### LINKED ISSUE #327: Istio manager cannot list of create k8s TPR when RBAC enabled
Seeing the following when running through the installation tutorial 
I've run: 
```
kubectl apply -f install/kubernetes/istio-rbac-beta.yaml
kubectl apply -f install/kubernetes/istio-auth.yaml
```

* k8s 1.6.4
* RBAC enabled

```
kubectl logs istio-manager-2910860705-c3zsh -c apiserver
I0526 03:10:04.952228       1 client.go:205] TPR "IstioConfig" is not ready (User "system:serviceaccount:default:istio-manager-service-account" cannot list istioconfigs.istio.io at the cluster scope. (get IstioConfigs.istio.io)). Waiting...
I0526 03:10:05.955367       1 client.go:205] TPR "IstioConfig" is not ready (User "system:serviceaccount:default:istio-manager-service-account" cannot list istioconfigs.istio.io at the cluster scope. (get IstioConfigs.istio.io)). Waiting...
Error: 2 errors occurred:
* failed to register Third-Party Resources. User "system:serviceaccount:default:istio-manager-service-account" cannot get thirdpartyresources.extensions at the cluster scope. (get thirdpartyresources.extensions istio-config.istio.io)
* failed to register Third-Party Resources. Failed to create all TPRs
```

The issue appears to be that the manager services are doing get at the cluster scope and we have a RoleBinding which is namespaced. Either you change the code to do a scoped get on namespace or change the following to a ClusterRoleBinding and add the namespace to the subject.

```
# Grant permissions to the Manager/discovery.
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: istio-manager-admin-role-binding
subjects:
- kind: ServiceAccount
  name: istio-manager-service-account
roleRef:
  kind: ClusterRole
  name: istio-manager
  apiGroup: rbac.authorization.k8s.io
```

### PR DESCRIPTION
**Pull request by lachie83**
Fixes #327

### CODE REVIEW

**Comment by andraxylia** [comment]
Is the "ingresses/status" needed for beta? The ingress status gets updated for me without this.

**Comment by lachie83** [comment]
I'm still seeing it under beta -- See comment in #327

**Comment by andraxylia** [comment]
This would limit to "default" namespace.

File: install/kubernetes/istio-rbac-beta.yaml [hunk 1]
```diff
@@ -3,6 +3,9 @@
 # Also helps to enable logging on apiserver 'wrap' to see the URLs.
 # Each RBAC deny needs to be mapped into a rule for the role.
 # If using minikube, start with '--extra-config=apiserver.Authorization.Mode=RBAC'
+#
+# NOTE: If deploying istio to a namespace other than 'default' then change the
+# ClusterRoleBinding namspace target appropriately.
```
Comment (costinm) [inline install/kubernetes/istio-rbac-beta.yaml]: I don't think we can ask users to edit the file - can you use RoleBinding instead ?

Comment (lachie83) [inline install/kubernetes/istio-rbac-beta.yaml]: We can if we scope the manager scope it's TPR list request to namespace only see initial logs in #327

**Comment by lachie83** [comment]
BTW - I didn't do anything here but bring the alpha and beta rbac rules that were already in the repo into sync. This isn't adding any new functionality. It seems folks have been testing against alpha and things were broken in beta.

**Comment by andraxylia** [comment]
The comment  #327 does not justify "ingresses/status", it's has the same cause as the main issue. The symptom in rbac alpha was the when you run "kubectl get ingress" in a cloud provider, there was no address: istio/istio#317 .

**Comment by lachie83** [comment]
You are absolutely correct @andraxylia. My bad -- the log message can be found here -- https://github.com/istio/istio/issues/329

**Comment by andraxylia** [comment]
Ok, good to know this is still the case with beta. Since I no longer hit the issue, I wrongly presumed the "status" got fixed to be part of "ingress".

**Comment by costinm** [comment]
I think the only problem with this PR is the namespace - if you can use RoleBinding instead of ClusterRoleBinding it would be great. Both grant the same ClusterRole - the main difference is that the 
former grants role to the service account in the current namespace, doesn't hardcode a specific namespace in the yaml.

**Comment by lachie83** [comment]
@costinm Agreed. I think that would just need a got change to scope the list of the TPRs then we are GTG with not doing this. Just to reiterate. It's a carbon copy of the alpha rbac manifests to the beta. I haven't changed anything other than the API version

File: install/kubernetes/istio-rbac-beta.yaml (hunk 1 above)
Comment (seanknox) [inline install/kubernetes/istio-rbac-beta.yaml]: @costinm the Kubernetes default is that service accounts are created in the `default` namespace. For the vast majority of users, they won't need to edit anything. 

If they want to run istio in a different namespace however, then they'll need ensure the namespace of the service account used by `ClusterRoleBinding` matches.

I think part of the confusion is that `ClusterRoleBinding` is scoped to the entire cluster—not a namespace—while it depends on a service account that is scoped to a namespace.

**Comment by rshriram** [comment]
@lachie83 This seems to be a regression from 1.5. I am happy to merge this but our release process is stalled as our repos are undergoing some restructuring. So to unblock folks who are facing this issue, I am willing to manually fix this in the 1.6 release artifact directly (without triggering Jenkins). (I don't want a 1.6.2 release without corresponding versions of pilot, mixer, etc).

But I need to know what exactly needs to be done. And couple of folks have to confirm that this fixed version works on kube 1.6, 1.5.2,1.5.3. 

cc @todkap @craigyam @andraxylia


The typo in your last comment is making it hard to understand what is the exact solution. RoleBinding or ClusterRoleBinding with regard to comments from @costinm.

**Comment by todkap** [comment]
I tried to do some validation of the updated istio-rbac-beta.yaml on Istio 0.1.6 and still hitting issues with the patch.  Please let me know if I can provide additional info.

> todkapmcbookpro:istio-0.1.5 todd$ istioctl version
> istioctl version:
> 
> Version: 0.1.6
> GitRevision: dab2033
> GitBranch: release-0.1
> User: jenkins@ubuntu-16-04-build-12ac793f80be71
> GolangVersion: go1.8
> KubeInjectHub: docker.io/istio
> KubeInjectTag: 0.1
> 
> 
> apiserver version:
> 
> Error: the server could not find the requested resource

[istio_ingress_failure.txt](https://github.com/istio/istio/files/1106576/istio_ingress_failure.txt)

**Comment by ldemailly** [comment]
we need to reproduce this issue, then fix it in the 0.1 branch - but not blindly if we don't have a repro first

cc @frankbu

**Comment by lachie83** [comment]
@rshriram apologies for the confusion. I've updated my comment. Hopefully that helps. I don't think that this is a regression but suspect that it was tested in a k8s 1.5 cluster which I believe uses the alpha RBAC APIs which already included the changes that are in this PR. When I used this on a 1.6 cluster the beta RBAC APIs are used which is where I could repro quickly. LMK if you need more details on the repro.

**Comment by ldemailly** [comment]
0.1.6 and beta rbac (1.6 k8s) works for me so yes I'd like to understand what is different in your setup

**Comment by todkap** [comment]
@ldemailly   I am more than willing to help provide details on my deployment.  Please let me know if I can provide output from any kubectl or iostoctl commands.

**Comment by todkap** [comment]
Quick update.   I was able to refocus back on debugging my deployment and found that I had some environment path issues and was pointing to the incorrect istio-rbac-beta.yaml (still pointing at the non-patched one).  When I used the one from the pull request, I saw a new set of errors (this time from the pilot).
` kubectl log istio-pilot-1836659236-67jv`

> I0628 20:38:47.695075       1 client.go:205] TPR "IstioConfig" is not ready (User "system:serviceaccount:default:istio-pilot-service-account" cannot list istioconfigs.istio.io at the cluster scope. (get IstioConfigs.istio.io)). Waiting...
> Error: 2 errors occurred:
> 
> * failed to register Third-Party Resources. User "system:serviceaccount:default:istio-pilot-service-account" cannot get thirdpartyresources.extensions at the cluster scope. (get thirdpartyresources.extensions istio-config.istio.io)
> * failed to register Third-Party Resources. Failed to create all TPRs
> Usage:
>   pilot apiserver [flags]
> 
> Flags:
>       --port int   Config API service port (default 8081)
> 
> Global Flags:
>       --domainSuffix string              Kubernetes DNS domain suffix (default "cluster.local")
>       --kubeconfig string                Use a Kubernetes configuration file instead of in-cluster configuration
>       --log_backtrace_at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
>       --meshConfig string                ConfigMap name for Istio mesh configuration, config key should be "mesh" (default "istio")
>   -n, --namespace string                 Select a namespace for the controller loop. If not set, uses ${POD_NAMESPACE} environment variable
>       --resync duration                  Controller resync interval (default 1s)
>   -v, --v Level                          log level for V logs (default 0)
>       --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
> 
> E0628 20:38:48.740562       1 main.go:247] 2 errors occurred:
> 
> * failed to register Third-Party Resources. User "system:serviceaccount:default:istio-pilot-service-account" cannot get thirdpartyresources.extensions at the cluster scope. (get thirdpartyresources.extensions istio-config.istio.io)
> * failed to register Third-Party Resources. Failed to create all TPRs

I introduced a new config element in the rbac to try to move past this.

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: istio-pilot
spec:
  replicas: 1
  template:
    metadata:
      annotations:
        alpha.istio.io/sidecar: ignore
      labels:
        istio: pilot
    spec:
      serviceAccountName: istio-pilot-service-account
      containers:
      - name: discovery
        image: docker.io/istio/pilot:0.1.6
        imagePullPolicy: Always
        args: ["discovery", "-v", "2"]
        ports:
        - containerPort: 8080
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
      - name: apiserver
        image: docker.io/istio/pilot:0.1.6
        imagePullPolicy: Always
        args: ["apiserver", "-v", "2"]
        ports:
        - containerPort: 8081
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
---
```

Now I am seeing issues with timeouts from the Ingress when I try to route through that.  I can directly hit the Node app the Ingress is fronting.

> todkapmcbookpro:istio-0.1.5 todd$ kubectl log istio-ingress-2247081378-rr3kz
> W0628 16:56:01.650785   21226 cmd.go:361] log is DEPRECATED and will be removed in a future version. Use logs instead.
> I0628 20:27:18.097800       1 client.go:171] Resource already exists: "istio-config.istio.io"
> I0628 20:27:18.098075       1 client.go:195] Checking for TPR resources
> I0628 20:27:18.099851       1 main.go:86] flags (main.args) {
>  kubeconfig: (string) "",
>  meshConfig: (string) (len=5) "istio",
>  ipAddress: (string) "",
>  podName: (string) "",
>  passthrough: ([]int) <nil>,
>  apiserverPort: (int) 8081,
>  controllerOptions: (kube.ControllerOptions) {
>   Namespace: (string) (len=7) "default",
>   ResyncPeriod: (time.Duration) 1s,
>   DomainSuffix: (string) (len=13) "cluster.local"
>  },
>  discoveryOptions: (envoy.DiscoveryServiceOptions) {
>   Port: (int) 8080,
>   EnableProfiling: (bool) true,
>   EnableCaching: (bool) true
>  }
> }
> I0628 20:27:18.102986       1 main.go:94] mesh configuration (*istio_proxy_v1_config.ProxyMeshConfig)(0xc420085970)(egress_proxy_address:"istio-egress:80" discovery_address:"istio-pilot:8080" mixer_address:"istio-mixer:9091" zipkin_address:"zipkin:9411" proxy_listen_port:15001 proxy_admin_port:15000 drain_duration:<seconds:2 > parent_shutdown_duration:<seconds:3 > istio_service_cluster:"istio-proxy" discovery_refresh_delay:<seconds:1 > connect_timeout:<seconds:1 > ingress_class:"istio" ingress_service:"istio-ingress" ingress_controller_mode:STRICT auth_policy:MUTUAL_TLS auth_certs_path:"/etc/certs" )
> I0628 20:27:18.109105       1 agent.go:164] Starting proxy agent
> I0628 20:27:18.109165       1 agent.go:182] Received new config, resetting budget
> I0628 20:27:18.109185       1 agent.go:247] Reconciling configuration (budget 10)
> I0628 20:27:18.109237       1 agent.go:267] Epoch 0 starting
> I0628 20:27:18.109259       1 config.go:42] writing configuration to /etc/envoy/envoy-rev0.json
> {
>   "listeners": [
>     {
>       "address": "tcp://0.0.0.0:80",
>       "filters": [
>         {
>           "type": "read",
>           "name": "http_connection_manager",
>           "config": {
>             "codec_type": "auto",
>             "stat_prefix": "http",
>             "generate_request_id": true,
>             "use_remote_address": true,
>             "tracing": {
>               "operation_name": "ingress"
>             },
>             "rds": {
>               "cluster": "rds",
>               "route_config_name": "80",
>               "refresh_delay_ms": 1000
>             },
>             "filters": [
>               {
>                 "type": "decoder",
>                 "name": "router",
>                 "config": {}
>               }
>             ],
>             "access_log": [
>               {
>                 "path": "/dev/stdout"
>               }
>             ]
>           }
>         }
>       ],
>       "bind_to_port": true
>     }
>   ],
>   "admin": {
>     "access_log_path": "/dev/stdout",
>     "address": "tcp://0.0.0.0:15000"
>   },
>   "cluster_manager": {
>     "clusters": [
>       {
>         "name": "zipkin",
>         "connect_timeout_ms": 1000,
>         "type": "strict_dns",
>         "lb_type": "round_robin",
>         "hosts": [
>           {
>             "url": "tcp://zipkin:9411"
>           }
>         ]
>       },
>       {
>         "name": "rds",
>         "connect_timeout_ms": 1000,
>         "type": "strict_dns",
>         "lb_type": "round_robin",
>         "hosts": [
>           {
>             "url": "tcp://istio-pilot:8080"
>           }
>         ]
>       }
>     ],
>     "sds": {
>       "cluster": {
>         "name": "sds",
>         "connect_timeout_ms": 1000,
>         "type": "strict_dns",
>         "lb_type": "round_robin",
>         "hosts": [
>           {
>             "url": "tcp://istio-pilot:8080"
>           }
>         ]
>       },
>       "refresh_delay_ms": 1000
>     },
>     "cds": {
>       "cluster": {
>         "name": "cds",
>         "connect_timeout_ms": 1000,
>         "type": "strict_dns",
>         "lb_type": "round_robin",
>         "hosts": [
>           {
>             "url": "tcp://istio-pilot:8080"
>           }
>         ]
>       },
>       "refresh_delay_ms": 1000
>     }
>   },
>   "tracing": {
>     "http": {
>       "driver": {
>         "type": "zipkin",
>         "config": {
>           "collector_cluster": "zipkin",
>           "collector_endpoint": "/api/v1/spans"
>         }
>       }
>     }
>   }
> }I0628 20:27:18.110813       1 watcher.go:154] Envoy command: [-c /etc/envoy/envoy-rev0.json --restart-epoch 0 --drain-time-s 2 --parent-shutdown-time-s 3 --service-cluster istio-proxy --service-node ingress]
> I0628 20:27:18.141524       1 leaderelection.go:203] attempting to acquire leader lease...
> [2017-06-28 20:27:18.149][9][warning][main] initializing epoch 0 (hot restart version=8.2490552)
> [2017-06-28 20:27:18.164][9][warning][main] starting main dispatch loop
> I0628 20:27:18.178514       1 leaderelection.go:213] successfully acquired lease default/istio-ingress-controller-leader-istio
> I0628 20:27:18.179617       1 status.go:167] new leader elected ()
> I0628 20:27:18.179687       1 status.go:169] I am the new status update leader
> W0628 20:27:19.141471       1 ingress.go:86] failed to fetch http://istio-pilot:8080/v1alpha/secret/istio-proxy/ingress Get http://istio-pilot:8080/v1alpha/secret/istio-proxy/ingress: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
> 

I will continue debugging.   Not sure if the pilot addition I made was the correct one(based it upon the istio-ingress-service-account.

**Comment by lachie83** [comment]
Just updated to include the rename from manager to pilot

**Comment by ldemailly** [comment]
I'm pretty sure some of those changes are already made (manager->pilot specifically) - can you rebase (either on release-0.1 or on master) ?

**Comment by lachie83** [comment]
rebase master

**Comment by rshriram** [comment]
Merge conflicts. Please fix and we can merge after tests pass.

**Comment by lachie83** [comment]
Looks like @andraxylia commited changes 18 hours ago that make most of this PR now irrelevant. There's one line left to change.

**Comment by andraxylia** [comment]
@lachie83 : I had to fix RBAC because it was not working anymore for me with the GKE 1.6.6 cluster
I forced merged the branch because it kept getting outdated, there was one more change needed. Please keep in mind we do not have a robot that merges automatically, so it's a good idea to merge right after you get the email about tests passing and the status is green, otherwise somebody else will commit again. If anybody else rebases your branch, you run into issues with the merge, so best is to keep an eye on the email. (hopefully we will get a robot soon). 
Also be aware this ties RBAC to the service account in the default namespace. If using Istio in another namespace, change the "namespace" line in the ClusterRoleBinding. I submitted a change to docs for this.

**Pull request merged**